manim/utils/decorators.py

```diff
@@ -0,0 +1,454 @@
+            params="a, b, c",
+            since="v0.2",
+            until="v0.4",
+            message="The letters x, y, z are cooler."
+        )
+        def foo(**kwargs):
+            pass
+
+        foo(a=2)
+        # WARNING  The parameter a of function foo has been deprecated since v0.2 and is expected to be deleted after v0.4. The letters x, y, z are cooler.
+
+    Basic parameter redirection::
+
+        @deprecated_params(redirections=[
+            #Two ways to redirect one parameter to another:
+            ("old_param", "new_param"),
+            lambda old_param2: {"new_param22": old_param2}
+        ])
+        def foo(**kwargs):
+            return kwargs
+
+        foo(x=1, old_param=2)
+        # WARNING  The parameter old_param of function foo has been deprecated and may be deleted in a future version.
+        # returns {"x": 1, "new_param": 2}
+
+    Redirecting using a calculated value::
+
+        @deprecated_params(redirections=[
+            lambda runtime_in_ms: {"run_time": runtime_in_ms / 1000}
+        ])
+        def foo(**kwargs):
+            return kwargs
+
+        foo(runtime_in_ms=500)
+        # WARNING  The parameter runtime_in_ms of function foo has been deprecated and may be deleted in a future version.
+        # returns {"run_time": 0.5}
+
+    Redirecting multiple parameter values to one::
+
+        @deprecated_params(redirections=[
+            lambda buff_x=1, buff_y=1: {"buff": (buff_x, buff_y)}
+        ])
+        def foo(**kwargs):
+            return kwargs
+
+        foo(buff_x=2)
+        # WARNING  The parameter buff_x of function foo has been deprecated and may be deleted in a future version.
+        # returns {"buff": (2, 1)}
+
+    Redirect one parameter to multiple::
+
+        @deprecated_params(redirections=[
+            lambda buff=1: {"buff_x": buff[0], "buff_y": buff[1]} if isinstance(buff, tuple)
+                    else {"buff_x": buff,    "buff_y": buff}
+        ])
+        def foo(**kwargs):
+            return kwargs
+
+        foo(buff=0)
+        foo(buff=(1,2))
+        # WARNING  The parameter buff of function foo has been deprecated and may be deleted in a future version.
+        # returns {"buff_x": 0, buff_y: 0}
+
+        # WARNING  The parameter buff of function foo has been deprecated and may be deleted in a future version.
+        # returns {"buff_x": 1, buff_y: 2}
+
+
+    """
+    # Check if decorator is used without parenthesis
+    if callable(params):
+        raise ValueError("deprecate_parameters requires arguments to be specified.")
+
+    # Construct params list
+    params = re.split("[,\s]+", params) if isinstance(params, str) else list(params)
+
+    # Add params which are only implicitly given via redirections
+    for redirector in redirections:
+        if isinstance(redirector, tuple):
+            params.append(redirector[0])
+        else:
+            params.extend(redirector.__code__.co_varnames)
+    params = list(set(params))
```

Comment:
isn't `params` already a list. why is it required?

Reply:
It is required, because after adding redirected parameters to it it may contain duplicates.